### PR TITLE
Complete Iteration 2

### PR DESCRIPTION
**Pull request by sergejs-hramkovs**
Iteration 2 Requirements 
1. Add information to represent count of iteration and count of live cells 
2. Add possibility to save information to file and restore it on application start 
3. Possibility to stop application in any moment

Requirements 1 and 3 were completed during the Iteration 1.

### CODE REVIEW

File: GameOfLife/GameOfLife/Engine.cs [hunk 1]
```diff
@@ -7,15 +7,21 @@ public class Engine
...
+        private Tuple<string[,], int, int, int> _renderReturnValues;
 
         /// <summary>
         /// Initiate field size choice.
         /// </summary>
         public void Start()
         {
-            string fieldSizeChoice;
-            Console.WriteLine("Welcome to the Game of Life!");
+            Console.SetWindowSize(170, 55);
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: why that specific size?

File: GameOfLife/GameOfLife/Engine.cs [hunk 2]
```diff
@@ -25,78 +31,123 @@ public void Start()
                     Console.WriteLine("Wrong Input!");
                     _wrongInput = false;
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: If you write wrong input, then why is the variable _wrongInput set to false?

File: GameOfLife/GameOfLife/Engine.cs [hunk 3]
```diff
@@ -25,78 +31,123 @@ public void Start()
...
-                                Console.Write("\nEnter the width of the field: ");
-                                if (int.TryParse(Console.ReadLine(), out _width) && _width > 0)
+                                if (_wrongInput)
                                 {
-                                    break;
+                                    Console.Clear();
+                                    Console.WriteLine("Wrong Input!");
+                                    _wrongInput = false;
+                                }
+                                Console.Write("\nEnter the length of the field: ");
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: Make all the string as constants, you can do that in a different static class only for constants.

File: GameOfLife/GameOfLife/Engine.cs [hunk 4]
```diff
@@ -7,15 +7,21 @@ public class Engine
...
+        private bool _loaded = false;
+        private bool _gliderGunMode = false;
+        private bool _resetGeneration = false;
+        private File _file;
+        private Tuple<string[,], int, int, int> _renderReturnValues;
 
         /// <summary>
         /// Initiate field size choice.
         /// </summary>
         public void Start()
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: This method is to big, need to simplify it.

File: GameOfLife/GameOfLife/Engine.cs [hunk 5]
```diff
@@ -25,78 +31,123 @@ public void Start()
...
+                            _length = 20;
+                            _width = 20;
+                            break;
+
+                        case ConsoleKey.D5:
+                            _length = 75;
+                            _width = 40;
+                            break;
+
+                        case ConsoleKey.D6:
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: the D6 logic can easily be brought out.

File: GameOfLife/GameOfLife/Engine.cs [hunk 6]
```diff
@@ -25,78 +31,123 @@ public void Start()
...
+                        case ConsoleKey.Escape:
+                            Environment.Exit(0);
+                            break;
+
+                        default:
+                            _length = 10;
+                            _width = 10;
+                            break;
+                    }
+                    if (_fieldSizeChoice.Key == ConsoleKey.D1 || _fieldSizeChoice.Key == ConsoleKey.D2 ||
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: you can add a variable in the cases where these specific characters are pressed and then set it to true and only check that one variable in the if.

File: GameOfLife/GameOfLife/Engine.cs [hunk 7]
```diff
@@ -25,78 +31,123 @@ public void Start()
...
+                    }
+                    else
+                    {
+                        if (_fieldSizeChoice.Key != ConsoleKey.G)
+                        {
+                            _wrongInput = true;
+                        }          
+                    }
                 }
                 else
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: this else block logic can be made a separate method

File: GameOfLife/GameOfLife/Engine.cs [hunk 8]
```diff
@@ -147,7 +198,28 @@ private void Pause(ConsoleKeyInfo keyPressed)
...
+                }             
+                else if (_saveKey.Key == ConsoleKey.Escape)
+                {
+                    Environment.Exit(0);
+                }
+                else if (_saveKey.Key == ConsoleKey.R)
+                {
+                    Restart();
+                }
+                else
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: why not use switch here?

File: GameOfLife/GameOfLife/Engine.cs [hunk 9]
```diff
@@ -156,40 +228,51 @@ private void Pause(ConsoleKeyInfo keyPressed)
         /// </summary>
         public void Run()
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: In general would be nicer to have method names more than just save or run, be more specific, save game run game etc.

File: GameOfLife/GameOfLife/Engine.cs [hunk 10]
```diff
@@ -156,40 +228,51 @@ private void Pause(ConsoleKeyInfo keyPressed)
...
+                Environment.Exit(0);
+            }
         }
 
         /// <summary>
         /// Method to count the current number of alive cells on the field.
         /// </summary>
+        /// <param name="gameField">An array of the game field cells.</param>
         /// <returns>Returns the number of alive cells currently in the gamefield array.</returns>
         public int CountAlive(string[,] gameField)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: Count alive what?

File: GameOfLife/GameOfLife/Engine.cs [hunk 11]
```diff
@@ -156,40 +228,51 @@ private void Pause(ConsoleKeyInfo keyPressed)
...
         {
             int aliveCellCount = 0;
 
             for (int i = 0; i < gameField.GetLength(0); i++)
             {
-                for (int j = 0; j < gameField.GetLength(0); j++)
+                for (int j = 0; j < gameField.GetLength(1); j++)
                 {
                     if (gameField[i, j] == "X")
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: X as const

File: GameOfLife/GameOfLife/Engine.cs [hunk 12]
```diff
@@ -203,14 +286,15 @@ public int CountAlive(string[,] gameField)
         /// <summary>
         /// Method to count the current number of dead cells on the field.
         /// </summary>
+        /// <param name="gameField">An array of the game field cells.</param>
         /// <returns>Returns the number of dead cells currently in the gamefield array.</returns>
         public int CountDead(string[,] gameField)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Engine.cs]: Why is this needed, and if we need count from all the cells from field that are not alive, why not take the cell count and subtract the alive cell count?

File: GameOfLife/GameOfLife/Field.cs [hunk 13]
```diff
@@ -10,6 +10,7 @@ public class Field
         private int _coordinateY;
         private bool _wrongInput = false;
         private bool _stop = false;
+        private ConsoleKeyInfo _seedingChoice;
 
         public Field(int length, int width)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Field.cs]: If you create the field here, then there is no need for _fieldLength and _fieldWidthm as you have those values already in the array itself as size.

File: GameOfLife/GameOfLife/Field.cs [hunk 14]
```diff
@@ -23,74 +24,60 @@ public Field(int length, int width)
...
-        }
+        }     
 
         /// <summary>
         /// Method to choose how to seed the field - manually or automatically.
         /// </summary>
+        /// <param name="gliderGunMode">Parameter to show whether the glider gun mode is on.</param>
         /// <returns>Returns an array of a seeded gamefield.</returns>
-        public string[,] SeedField()
+        public string[,] SeedField(bool gliderGunMode)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Field.cs]: Maybe would be better to call it PopulateField instead of SeedField?

File: GameOfLife/GameOfLife/Field.cs [hunk 15]
```diff
@@ -176,41 +163,43 @@ public void DrawField(string[,] field)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Field.cs]: As I can't add a comment on randomSeeding, then I have a question, so in an most unlikely scenario, but if the random would always return the same coordinates, then if you would for example have that you need 5 alive cells, but the random gives the same coordinates all the time, or from the 5 times only two are not the same, then instead of 5 alive cells you would have 2 or 3, so looks kind of sketchy.

File: GameOfLife/GameOfLife/Field.cs [hunk 16]
```diff
@@ -176,41 +163,43 @@ public void DrawField(string[,] field)
...
-        /// </summary>
+        /// </summary
+        /// <param name="gliderGunMode">Parameter to show whether the glider gun mode is on.</param>
         /// <returns>Returns an array of a gamefield seeded with objects from the library.</returns>
-        private string[,] LibrarySeeding()
+        private string[,] LibrarySeeding(bool gliderGunMode)
         {
-            string inputPattern;
+            ConsoleKeyInfo libraryChoice;
             Library library = new Library(_fieldArray);
```
Comment (taurins) [inline GameOfLife/GameOfLife/Field.cs]: Inject the Library object.

File: GameOfLife/GameOfLife/Field.cs [hunk 17]
```diff
@@ -176,41 +163,43 @@ public void DrawField(string[,] field)
...
-                Console.WriteLine("2. Spawn a light-weight spaceship");
-                Console.WriteLine("3. Spawn a middle-weight spaceship");
-                Console.WriteLine("4. Spawn a heavy-weight spaceship");
-                Console.Write("\nChoice: ");
-                inputPattern = Console.ReadLine();
+                Render.LibraryMenuRender();
+                libraryChoice = Console.ReadKey(true);
 
-                switch (inputPattern)
+                switch (libraryChoice.Key)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Field.cs]: why are cases D1 to D4 the same? You can bring that code out in a separate method, as we don't want to repeat ourselves.

File: GameOfLife/GameOfLife/Field.cs [hunk 18]
```diff
@@ -263,7 +252,7 @@ public void DrawField(string[,] field)
         /// <returns>Returns "stop = true" if the process of entering coordinates was stopped. Returns false if there is wrong input.</returns>
         private bool EnterCoordinates()
         {
-            Console.WriteLine("\nTo stop seeding enter 'stop'");
+            Console.WriteLine("\n# To stop seeding enter 'stop'");
             Console.Write("\nEnter X coordinate: ");
             _inputCoordinate = Console.ReadLine();
             if (_inputCoordinate == "stop")
```
Comment (taurins) [inline GameOfLife/GameOfLife/Field.cs]: why do you have this? return _stop = true; on line 260

File: GameOfLife/GameOfLife/File.cs [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+            }
+            return _stringField;
+        }
+
+        /// <summary>
+        /// Method that takes the list of the game field cells and converts it to an array of the game field.
+        /// </summary>
+        /// <param name="inputList">List of the game field cells.</param>
+        /// <returns>Returns an array of the game field.</returns>
+        private string[,] ListToField(List<string> inputList)
```
Comment (taurins) [inline GameOfLife/GameOfLife/File.cs]: A more descriptive name please.

File: GameOfLife/GameOfLife/File.cs [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+            _writer.Close();
+        }
+
+        /// <summary>
+        /// Method to load the saved field from the file.
+        /// </summary>
+        /// <returns>Returns call to ListToField method, which returns an array of the gamefield.</returns>
+        public string[,] LoadFromFile()
+        {
+            _reader = new StreamReader(_filePath);
```
Comment (taurins) [inline GameOfLife/GameOfLife/File.cs]: what happens if the file has been misplaced or does not exist at all?

File: GameOfLife/GameOfLife/Iteration.cs [hunk 21]
```diff
@@ -152,5 +162,125 @@ private void CheckDead(string[,] field)
             _cellsToDie.Clear();
             return field;
         }
+
+        /// <summary>
+        /// Modification of CheckAlive method with the addition of dead borders.
+        /// </summary>
+        /// <param name="field">An array of a gamefield.</param>
+        private void CheckAliveDeadBorder(string[,] field)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Iteration.cs]: How to understand what does dead alive border mean?

File: GameOfLife/GameOfLife/Iteration.cs [hunk 22]
```diff
@@ -152,5 +162,125 @@ private void CheckDead(string[,] field)
...
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Modification of CheckDead method with the addition of dead borders.
+        /// </summary>
+        /// <param name="field">An array of a gamefield.</param>
+        private void CheckDeadDeadBorder(string[,] field)
```
Comment (taurins) [inline GameOfLife/GameOfLife/Iteration.cs]: what does dead dead mean?

File: GameOfLife/GameOfLife/Iteration.cs [hunk 23]
```diff
@@ -152,5 +162,125 @@ private void CheckDead(string[,] field)
...
+
+                        if (neigboursCountOfDead == 3 && i != field.GetLength(0) - 1 && j != field.GetLength(1) - 1)
+                        {
+                            _cellsToBeBorn.Add((i, j));
+                        }
+                        neigboursCountOfDead = 0;
+                    }
+                }
+            }
+        }
```
Comment (taurins) [inline GameOfLife/GameOfLife/Iteration.cs]: These four methods are almost identical, so instead of writing four different methods it would be better to write one and pass the needed values to that and get the needed result back, this is a very bad approach that you have here.

File: GameOfLife/GameOfLife/Render.cs [hunk 24]
```diff
@@ -1,58 +1,168 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
-
-namespace GameOfLife
+﻿namespace GameOfLife
 {
     public static class Render
```
Comment (taurins) [inline GameOfLife/GameOfLife/Render.cs]: why is this a static class?

File: GameOfLife/GameOfLife/Render.cs [hunk 25]
```diff
@@ -1,58 +1,168 @@
...
             gameField = inputField;
             field = new Field(length, width);
             iteration = new Iteration();
             engine = new Engine();
-            gameField = field.CreateField();
-            Console.Clear();
-            field.DrawField(gameField);
-            gameField = field.SeedField();
+
+            // This place needs to be had a look at.
```
Comment (taurins) [inline GameOfLife/GameOfLife/Render.cs]: these kind of comments shouldn't be in a PR

File: GameOfLife/GameOfLife/Render.cs [hunk 26]
```diff
@@ -1,58 +1,172 @@
...
-        public static void InitialRender(int length, int width, string[,] inputField)
+        /// <param name="loaded">Boolean parameter that represents whether the field was loaded from the file.</param>
+        /// <param name="gliderGunMode">Parameter to show whether the glider gun mode is on.</param>
+        public static void InitialRender(int length, int width, string[,] inputField, bool loaded, bool gliderGunMode)
         {
             gameField = inputField;
             field = new Field(length, width);
-            iteration = new Iteration();
+            rulesApplier = new RulesApplier();
             engine = new Engine();
```
Comment (taurins) [inline GameOfLife/GameOfLife/Render.cs]: need to inject class objects, not create new ones everywhere.

**Review by taurins — CHANGES_REQUESTED** [review]

File: GameOfLife/GameOfLife/Engine.cs (hunk 1 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: I just wanted to display a big game field without resizing the console window. I have chosen 75x40 field, and then I have come up with this window size. I can change it or remove it, if it's not a good practice.

File: GameOfLife/GameOfLife/Engine.cs (hunk 2 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: I wanted to reset the variable before the next input attempt. But still this code does not seem to work as I intended. I have re-made this place.

File: GameOfLife/GameOfLife/Engine.cs (hunk 3 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: Moved all the often used strings of phrases to a separate class.

File: GameOfLife/GameOfLife/Engine.cs (hunk 4 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: Simplified this method. Made two additional methods from the code from this method.

File: GameOfLife/GameOfLife/Engine.cs (hunk 5 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: Created a separate method for dealing with D6.

File: GameOfLife/GameOfLife/Engine.cs (hunk 6 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: Simplified that 'if' condition.

File: GameOfLife/GameOfLife/Engine.cs (hunk 7 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: I guess I have just taken care of this place, if I am not mistaken.

File: GameOfLife/GameOfLife/Engine.cs (hunk 8 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: Right, changed this if-else to switch. I have been migrating all the necessary initial if-else's to switch, might have missed some.

File: GameOfLife/GameOfLife/Engine.cs (hunk 9 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: Changed those names.

File: GameOfLife/GameOfLife/Engine.cs (hunk 10 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: Corrected this name and other similar.

File: GameOfLife/GameOfLife/Engine.cs (hunk 11 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: Moved Alive and Dead cell symbols to the static class with string constants.

File: GameOfLife/GameOfLife/Engine.cs (hunk 12 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Engine.cs]: I just thought that it would be nice to also see the number of dead cells. CountDead method seems to be absolutely unnecessary, changed it to simple subtraction.

File: GameOfLife/GameOfLife/Field.cs (hunk 13 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Field.cs]: Removed _fieldLength and __fieldWidth, and modified the code correspondingly.

File: GameOfLife/GameOfLife/Field.cs (hunk 14 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Field.cs]: Yes, sound reasonable. Change SeedField to PopulateFiled. Can I  leave '...Seeding' in the ManualSeeding, RandomSeeding and LibrarySeeding methods, I think it sounds appropriate enough.

File: GameOfLife/GameOfLife/Field.cs (hunk 15 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Field.cs]: Well, I think it is quite unlikely that randomizer will frequently return the same field. The amount of alive cells is random, cell's X and Y coordinates are random. The sole purpose of this type of seeding is to populate the field quickly, without entering anything manually. If a user wants something specific, there are other types of seeding. 

But what would be your suggestion to change in this method? I do not really get what can and should be changed.

File: GameOfLife/GameOfLife/Field.cs (hunk 16 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Field.cs]: Injected it.

File: GameOfLife/GameOfLife/Field.cs (hunk 17 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Field.cs]: Created a separate method for that logic.

File: GameOfLife/GameOfLife/Field.cs (hunk 18 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Field.cs]: Well, I wanted to stop the coordinates entering process, but it seems that my that implementation didn't work as intended. I have changed the code a bit, now if the user types in 'stop', the entering process stops. If the user enters anything else than a number or 'stop', the user also get 'Wrong Input!' message. Hope this new implementation is acceptable.

File: GameOfLife/GameOfLife/File.cs (hunk 19 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/File.cs]: Changed the names.

File: GameOfLife/GameOfLife/File.cs (hunk 20 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/File.cs]: An exception will be thrown. Added check for this situation.

File: GameOfLife/GameOfLife/Iteration.cs (hunk 21 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Iteration.cs]: At the time of writing this, it seemed understandable enough for me. But now I see, that it's not a very good name (as well as other similar). Changed these names.

File: GameOfLife/GameOfLife/Iteration.cs (hunk 22 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Iteration.cs]: Changed this name as well.

File: GameOfLife/GameOfLife/Field.cs (hunk 15 above)
Comment (taurins) [reply GameOfLife/GameOfLife/Field.cs]: One solution to the probability of duplicate coordinates is to add a check if the new generated coordinate is already set to spawn something.

**Review by taurins — APPROVED** [review]

File: GameOfLife/GameOfLife/Iteration.cs (hunk 23 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Iteration.cs]: Simplified and removed some methods from this class.

File: GameOfLife/GameOfLife/Render.cs (hunk 24 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Render.cs]: I thought that I didn't want to instantiate objects of this class. I have later made it non static.

File: GameOfLife/GameOfLife/Render.cs (hunk 25 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Render.cs]: Removed that comment.

File: GameOfLife/GameOfLife/Render.cs (hunk 26 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Render.cs]: Removed object creation, added object injection.

File: GameOfLife/GameOfLife/Field.cs (hunk 15 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Field.cs]: Oh, now I get what you mean. I misunderstood you on Friday. 
Added function repeating, if new random cell is already inhabited by an alive cell.

**Review by taurins — APPROVED** [review]

File: GameOfLife/GameOfLife/Render.cs (hunk 26 above)
Comment (taurins) [reply GameOfLife/GameOfLife/Render.cs]: k

File: GameOfLife/GameOfLife/Field.cs (hunk 15 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Field.cs]: That my idea caused occasional endless loops. Changed that place a bit, now it seems to work fine.

File: GameOfLife/GameOfLife/FileIO.cs [hunk 27]
```diff
@@ -0,0 +1,137 @@
+﻿using GameOfLife.Interfaces;
+using GameOfLife.Models;
+
+namespace GameOfLife
+{
+    public class FileIO : IFileIO
+    {
+        private string _filePath = @"C:\Users\sergejs.hramkovs\OneDrive - Accenture\Documents\field.txt";
```
Comment (rsolov) [inline GameOfLife/GameOfLife/FileIO.cs]: choose folder so others can run it too

Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/FileIO.cs]: Changed the filepath and added directory creation if it doesn't exist.

File: GameOfLife/GameOfLife/Services/FieldOperations.cs [hunk 28]
```diff
@@ -0,0 +1,221 @@
+﻿using GameOfLife.Interfaces;
+using GameOfLife.Models;
+using static GameOfLife.StringConstantsModel;
+
+namespace GameOfLife
+{
+    public class FieldOperations : IFieldOperations
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/FieldOperations.cs:10]: Add comments for class and constructor

File: GameOfLife/GameOfLife/Services/FieldOperations.cs [hunk 29]
```diff
@@ -0,0 +1,221 @@
...
+        {
+            get => _coordinateX;
+            set => _coordinateX = value;
+        }
+        public int CoordinateY
+        {
+            get => _coordinateY;
+            set => _coordinateY = value;
+        }
+        public bool Stop
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/FieldOperations.cs]: You can have implicit properties:
public bool Stop { get; set; }

Stop is too generic name. Make it more specific.

File: GameOfLife/GameOfLife/Services/FieldOperations.cs [hunk 30]
```diff
@@ -0,0 +1,221 @@
...
+                            break;
+
+                        case 2:
+                            _library.SpawnSimkinGliderGun(gameField, 0, 16);
+                            break;
+                    }
+                    Console.Clear();
+                    return gameField;
+                }
+                if (_inputProcessor.WrongInput)
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/FieldOperations.cs]: Here and in other places, add empty line after closing block.

File: GameOfLife/GameOfLife/Services/FileIO.cs [hunk 31]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        /// <summary>
+        /// Method that takes the list of the game field cells and converts it to an array of the game field.
+        /// </summary>
+        /// <param name="inputList">List of the game field cells.</param>
+        /// <returns>Returns an instance of the GameFieldModel class.</returns>
+        private GameFieldModel ConvertListOfRowsToGameField(List<string> inputList)
+        {
+            int x = 0;
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/FileIO.cs]: More descriptive names needed for variables x and y.

File: GameOfLife/GameOfLife/Services/FileIO.cs [hunk 32]
```diff
@@ -0,0 +1,228 @@
...
+            foreach (char character in inputList[0])
+            {
+                if (int.TryParse(character.ToString(), out int number))
+                {
+                    generationString += number.ToString();
+                }
+            }
+            gameField.Generation = int.Parse(generationString);
+
+            for (int i = 4; i < inputList.Count; i++)
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/FileIO.cs]: It is hard to read what i is. Make it more descriptive.

File: GameOfLife/GameOfLife/Services/RulesApplier.cs [hunk 33]
```diff
@@ -0,0 +1,138 @@
+﻿using GameOfLife.Interfaces;
+using GameOfLife.Models;
+using static GameOfLife.StringConstantsModel;
+
+namespace GameOfLife
+{
+    public class RulesApplier : IRulesApplier
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/RulesApplier.cs:10]: Add class comment

File: GameOfLife/GameOfLife/Services/RulesApplier.cs [hunk 34]
```diff
@@ -0,0 +1,138 @@
...
+        private List<(int x, int y)> _cellsToBeBorn = new List<(int x, int y)>();
+        private int neighboursCount;
+
+        /// <summary>
+        /// Method to determine which cells die and which are reborn, judging by the number of alive neighbours.
+        /// </summary>
+        /// <param name="gameField">An instance of the GameFieldModel class that stores the game field and its properties.</param>
+        /// <param name="disableWrappingAroundField">Parameter that shows if field's wrapping around is enabled.</param>
+        public void DetermineCellsDestiny(GameFieldModel gameField, bool disableWrappingAroundField)
+        {
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/RulesApplier.cs:22]: Logic for this method is not easy to read. Split it into three methods. This stays as master method which depending on disableWrappingAroundField calls either method for wrapped world or other for non wrapped world.

File: GameOfLife/GameOfLife/Services/RulesApplier.cs [hunk 35]
```diff
@@ -0,0 +1,138 @@
...
+
+        /// <summary>
+        /// Method to count each cell's alive and dead neighbours to later determine which will die and which will be reborn.
+        /// </summary>
+        /// <param name="gameField">An instance of the GameFieldModel class that stores the game field and its properties.</param>
+        /// <param name="x">Horizontal coordinate of a cell which neighbours are being counted.</param>
+        /// <param name="y">Vertical coordinate of a cell which neighbours are being counted</param>
+        /// <param name="checkDeadCell">Parameter which shows whether the cell, which neighbours are being counted, is alive or dead.</param>
+        /// <returns>Returns the number of cell's neighbours.</returns>
+        private int CountNeighbourCells(GameFieldModel gameField, int x, int y, bool checkDeadCell)
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/RulesApplier.cs]: Parameters x and y should be replaced by descriptive ones.

File: GameOfLife/GameOfLife/Views/Render.cs [hunk 36]
```diff
@@ -0,0 +1,312 @@
...
+
+        /// <summary>
+        /// Method for rendering the UI for choosing which saved game file to load.
+        /// </summary>
+        /// <param name="numberOfFiles">The number of saved game files currently present in the folder.</param>
+        public void ChooseFileToLoadMenuRender(int numberOfFiles, string filePath, bool wrongInput)
+        {
+            Console.CursorVisible = true;
+            Console.Clear();
+            Console.WriteLine(" ### Choose which saved game to load ###");
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Views/Render.cs:283]: Code should not contain so called 'magic strings' or 'magic numbers'.
Text or number literals should be placed into resource file or in constant variable at the start of class.

File: GameOfLife/GameOfLife/Services/FieldOperations.cs (hunk 28 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FieldOperations.cs:10]: Added missing comments.

File: GameOfLife/GameOfLife/Services/FieldOperations.cs (hunk 29 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FieldOperations.cs]: Changed properties to implicit properties. Made 'Stop' more specific.

File: GameOfLife/GameOfLife/Services/FieldOperations.cs (hunk 30 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FieldOperations.cs]: I am not sure what the  'closing block' is, where exactly should I put empty lines? Between the 'if' blocks?

File: GameOfLife/GameOfLife/Controllers/InputController.cs [hunk 37]
```diff
@@ -9,35 +9,15 @@ namespace GameOfLife
...
-        {
-            get => _gameField;
-            set => _gameField = value;
-        }
-        public MultipleGamesModel MultipleGames
-        {
-            get => _multipleGames;
-            set => _multipleGames = value;
-        }
+        public bool WrongInput { get; set; } = false;
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Controllers/InputController.cs]: Assignment of false isn't needed. Bool type get's false by default.

File: GameOfLife/GameOfLife/Services/FileIO.cs (hunk 31 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FileIO.cs]: Changed the names of variables x and y.

File: GameOfLife/GameOfLife/Services/FileIO.cs [hunk 38]
```diff
@@ -13,36 +13,13 @@ public class FileIO : IFileIO
...
-        public bool NoSavedGames
-        {
-            get => _noSavedGames;
-            set => _noSavedGames = value;
-        }
-        public int NumberOfFiles
-        {
-            get => _numberOfFiles;
-        }
+        public string FilePath { get; } = @"C:\GameOfLife_SavedGames\";
```
Comment (GuntarsD) [inline GameOfLife/GameOfLife/Services/FileIO.cs]: Magic string => ' @"C:\GameOfLife_SavedGames\";'
Should be taken out of code. You can create static class for parameters and place data there.
One implementation shortcut/improvement is to put data in subfolder of folder where running application is located.
You can get that by:
AppDomain.CurrentDomain.BaseDirectory

File: GameOfLife/GameOfLife/Services/FieldOperations.cs (hunk 30 above)
Comment (GuntarsD) [reply GameOfLife/GameOfLife/Services/FieldOperations.cs]: In C#, a code block is a group of lines of code between curly braces {}.
After closing brace there should be empty line except block is followed by closing brace of enclosing block.

So:
image
On image red lines where no empty line, green line where empty line should be.

File: GameOfLife/GameOfLife/Services/FileIO.cs (hunk 31 above)
Comment (GuntarsD) [reply GameOfLife/GameOfLife/Services/FileIO.cs]: It is better but awkward.
May be
int horizontalX = 0;    =>   column or horizontalOffset or xCoordinate
and 
int verticalY = 0;    =>   row or verticalOffset or yCoordinate

File: GameOfLife/GameOfLife/Views/Render.cs (hunk 36 above)
Comment (GuntarsD) [reply GameOfLife/GameOfLife/Views/Render.cs:283]: Will discuss this personally

File: GameOfLife/GameOfLife/Services/FieldOperations.cs (hunk 30 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FieldOperations.cs]: So there should be no empty lines between 'if' and 'else? And also, should I add an empty line in such a situation?
image

Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FieldOperations.cs]: Added necessary empty lines.

File: GameOfLife/GameOfLife/Services/FileIO.cs (hunk 31 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FileIO.cs]: Yes, I agree. At first I changed x and y to horizontalCoordinate and verticalCoordinate, but then I thought that it was a bit too long and changed. 

xCoordinate and yCoordinate looks good, changed to such names.

File: GameOfLife/GameOfLife/Services/FileIO.cs (hunk 32 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FileIO.cs]: Changed the names of 'i' and 'j'.

File: GameOfLife/GameOfLife/Services/RulesApplier.cs (hunk 33 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/RulesApplier.cs:10]: Added class comments.

File: GameOfLife/GameOfLife/Services/RulesApplier.cs (hunk 34 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/RulesApplier.cs:22]: Split the method into three ones.

File: GameOfLife/GameOfLife/Services/RulesApplier.cs (hunk 35 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/RulesApplier.cs]: Changed these names.

File: GameOfLife/GameOfLife/Controllers/InputController.cs (hunk 37 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Controllers/InputController.cs]: Removed boolean false assignments.

**Review by GuntarsD — APPROVED** [review]

File: GameOfLife/GameOfLife/Services/FileIO.cs (hunk 38 above)
Comment (sergejs-hramkovs) [reply GameOfLife/GameOfLife/Services/FileIO.cs]: Made the location of the SavedGames folder relative to the project folder.
